File: ruby/complex-numbers/complex_numbers.rb
class ComplexNumber
  attr_reader :real, :imaginary

  def initialize(real, imaginary)
    @real = real
    @imaginary = imaginary
  end

  def ==(c)
    real == c.real && imaginary == c.imaginary
  end

  def *(c)
    self.class.new(
      real * c.real - imaginary * c.imaginary,
      real * c.imaginary + imaginary * c.real)
  end

  def /(c)
    self * c.conjugate * self.class.new(c.abs**-2, 0)
  end

  def +(c)
    self.class.new(
      real + c.real,
      imaginary + c.imaginary)
  end

  def -(c)
    self.class.new(
      real - c.real,
      imaginary - c.imaginary)
  end

  def abs
    Math.sqrt([real, imaginary].sum { |p| p**2 })
  end

  def conjugate
    self.class.new(
      real,
      -imaginary)
  end

  def exp
    self.class.new(
      Math.exp(real) * Math.cos(imaginary), 
      Math.exp(real) * Math.sin(imaginary))
  end
end

File: clojure/complex-numbers/src/complex_numbers.clj
(ns complex-numbers)


(def real first)
(def imaginary second)
(def add (partial map +))
(def sub (partial map -))

(defn- sq [c] (reduce + (map * c c)))

(defn abs [c] (Math/sqrt (sq c)))
(defn conjugate [c] (update c 1 -))

(defn- op [f g [a b] [c d]] [(f (* a c) (* b d)) (g (* b c) (* a d))])

(defn mul [c d] (op - + c d))

(defn div [c [x y :as d]]
  {:pre [(not (== 0 x y))]}
  (let [div #(/ % (double (sq d)))]
    (op (comp div +) (comp div -) c d)))
